### PR TITLE
sql: add version gate to regrole type

### PR DESCRIPTION
**Pull request by RichardJCai**
Without the version gate, in a mixed cluster setting, creating
a table on the new binary version with type REGROLE can cause
errors when inserting on the old verison.

Errors can also happen on schema changes where the old binary uses
the new type.

Release justification: GA blocker, without a version gate this can
cause panics.
Release note: None. Not in production.

### CODE REVIEW

File: pkg/BUILD.bazel
Comment (RichardJCai) [inline pkg/BUILD.bazel:2]: Looks like my bazel-generate messed up. Will fix.

File: pkg/sql/types/minimum_type_version.go
Comment (rafiss) [inline pkg/sql/types/minimum_type_version.go:23]: do we also need a check to make sure the type doesn't appear in TupleContents for tuples?

File: pkg/sql/create_table_test.go
Comment (rafiss) [inline pkg/sql/create_table_test.go:11]: should this be in the `types` package and the file named `pkg/sql/types/minimum_type_version_test.go`

File: pkg/clusterversion/cockroach_versions.go
Comment (rafiss) [inline pkg/clusterversion/cockroach_versions.go:478]: let's just get additional sign-off to make sure it's ok to backport this. i believe it is... but if not then i guess the way to unblock the release would be to revert the regrole change.

File: pkg/sql/logictest/testdata/logic_test/regrole_type_mixed
Comment (rafiss) [inline pkg/sql/logictest/testdata/logic_test/regrole_type_mixed:15]: maybe also add tests for arrays and tuples?

**Review by rafiss — CHANGES_REQUESTED** [review]

**Review by nvb — APPROVED** [review]
Thanks!

> Reviewed 10 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70531#-:-Mk9FBR71fQRcTs0tk8J:b-ucfknh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RichardJCai)

___
*[pkg/sql/types/minimum_type_version.go, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70531#-Mk9F0tGA8rlxEw3fWqG:-Mk9F0tGA8rlxEw3fWqH:bq8d6r2) ([raw file](https://github.com/cockroachdb/cockroach/blob/24d2ececdf64a184ce39068989864c6fdbbe85ff/pkg/sql/types/minimum_type_version.go#L16)):*
> ```Go
> 
> // minimumTypeUsageVersions defines the minimum version needed for a new
> // data type.
> ```

nit: consider adding a comment here advising people not to remove this map and the `IsTypeSupportedInVersion` function even if it ever drops to being empty temporarily.

**Comment by ajwerner** [comment]
Let's just piggy-back on some other version added in this cycle. I don't think we need to add a new one.

**Comment by RichardJCai** [comment]
> Let's just piggy-back on some other version added in this cycle. I don't think we need to add a new one.

Ack, does it make sense to piggy back off `MarkerDataKeysRegistry` here - the last version added.
https://github.com/cockroachdb/cockroach/blob/release-21.2/pkg/clusterversion/cockroach_versions.go#L473

Added in:
https://github.com/cockroachdb/cockroach/commit/f79c56d7940f3950feb64a2316839e0d230bb05a#diff-8f32466c40e6960b988c87d7e2453c01a591d420af123f0963959da8e580f2e4

**Review by ajwerner — COMMENTED** [review]
Fine by me. How about we add a const in the `sql` package that has a useful name and is equal to the `clusterversion` key you choose. That can be a good place to anchor the comment. Then the version check will look like it makes sense. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70531#-:-Mk9rS-S0tojidsqKuIH:b-8kkum2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)

File: pkg/sql/types/minimum_type_version.go
Comment (RichardJCai) [reply pkg/sql/types/minimum_type_version.go:29]: I don't think that's necessary because TupleContents can't be the type of a column.

File: pkg/sql/create_table_test.go
Comment (RichardJCai) [reply pkg/sql/create_table_test.go]: Done

File: pkg/sql/logictest/testdata/logic_test/regrole_type_mixed
Comment (RichardJCai) [reply pkg/sql/logictest/testdata/logic_test/regrole_type_mixed:21]: Added for arrays.

**Review by RichardJCai — COMMENTED** [review]
Added an alias for `MarkerDataKeysRegistry` in the `minimum_type_version` file.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70531#-:-MkDIMBp8iC1OLJLM5m_:b-y2qvrc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @rafiss, and @RichardJCai)

File: docs/generated/settings/settings.html
Comment (rafiss) [inline docs/generated/settings/settings.html:173]: i think these docs changes should be reverted now

**Review by rafiss — APPROVED** [review]
lgtm after fixing html docs!

File: pkg/sql/types/minimum_type_version.go
Comment (ajwerner) [inline pkg/sql/types/minimum_type_version.go:27]: Would it simplify the code if you returned an error rather than bool from this helper? 

I can see it both ways.

**Review by ajwerner — APPROVED** [review]
LGTM

File: pkg/sql/types/minimum_type_version.go
Comment (RichardJCai) [reply pkg/sql/types/minimum_type_version.go:27]: All the errors are the same so it would simplify the code in that sense but I prefer returning a bool over an error for these type of functions, I find that sometimes I come back to these type of functions that return errors and wish they returned a bool instead for flexibility.

Comment (ajwerner) [reply pkg/sql/types/minimum_type_version.go:27]: I feel that. I think there's a middle ground which is to create a helper to generate the error:

```go
package sql

func checkTypeIsSupported(ctx context.Context, settings *cluster.Settings, typ *types.T) error {
    version := settings.Version.ActiveVersionOrEmpty(ctx)
    if supported := types.IsTypeSupportedInVersion(version, typ); !supported {
        return pgerror.Newf(
            pgcode.FeatureNotSupported,
            "type %s is not supported until version upgrade is finalized",
            typ.SQLString(),
	)
    }
} 
```

Comment (RichardJCai) [reply pkg/sql/types/minimum_type_version.go:27]: Nice suggestion, done.

**Review by ajwerner — APPROVED** [review]
:lgtm_strong: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70531#-:-MkIrVaP0hjFPHyWloUD:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @nvanbenschoten, @rafiss, and @RichardJCai)

**Comment by RichardJCai** [comment]
TFTR!
bors r+

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70531)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/types/minimum_type_version.go, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70531#-Mk9F0tGA8rlxEw3fWqG:-Mx7_ruL00O2SfqOPQqV:b-aar5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/24d2ececdf64a184ce39068989864c6fdbbe85ff/pkg/sql/types/minimum_type_version.go#L16)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: consider adding a comment here advising people not to remove this map and the `IsTypeSupportedInVersion` function even if it ever drops to being empty temporarily.
</blockquote></details>

@richardjcai @nvanbenschoten Just stumbled across this code and wondering why we have this recommendation to not remove the map even it becomes empty? Is it so that in the future, when we will need a similar version gate, we already have the necessary infrastructure?

**Comment by RichardJCai** [comment]
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70531)_ status: :shipit: complete! 1 of 0 LGTMs obtained
> 
> _[pkg/sql/types/minimum_type_version.go, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70531#-Mk9F0tGA8rlxEw3fWqG:-Mx7_ruL00O2SfqOPQqV:b-aar5q) ([raw file](https://github.com/cockroachdb/cockroach/blob/24d2ececdf64a184ce39068989864c6fdbbe85ff/pkg/sql/types/minimum_type_version.go#L16)):_
> 
> Previously, nvanbenschoten (Nathan VanBenschoten) wrote…
> @RichardJCai @nvanbenschoten Just stumbled across this code and wondering why we have this recommendation to not remove the map even it becomes empty? Is it so that in the future, when we will need a similar version gate, we already have the necessary infrastructure?

Yeah that is exactly why, so we have the necessary infrastructure.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/70531)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/types/minimum_type_version.go, line 16 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/70531#-Mk9F0tGA8rlxEw3fWqG:-MxCMezS7-226WOjYZca:bxwl0n3) ([raw file](https://github.com/cockroachdb/cockroach/blob/24d2ececdf64a184ce39068989864c6fdbbe85ff/pkg/sql/types/minimum_type_version.go#L16)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Yeah that is exactly why, so we have the necessary infrastructure.
</blockquote></details>

I see, thanks. I'll probably remove it though in #77260 in order to break the dependency of `sql/types` on `roachpb` (via `clusterversion`). Do you have strong feelings about it?

**Pull request merged**
